pennylane/tape/tapes/tape.py

```diff
@@ -97,30 +96,24 @@ def expand_tape(tape, depth=1, stop_at=None, expand_measurements=False):
 
     new_tape = tape.__class__()
 
-    if expand_measurements:
-        # Check for observables acting on the same wire. If present, observables must be
-        # qubit-wise commuting Pauli words. In this case, the tape is expanded with joint
-        # rotations and the observables updated to the computational basis. Note that this
-        # expansion acts on the original tape in place.
-
-        obs = [m.obs for m in tape.measurements if m.obs is not None]
-        m_idx = [i for i, m in enumerate(tape.measurements) if m.obs is not None]
-        obs_wires = [wire for m in tape.measurements for wire in m.wires if m.obs is not None]
-
-        if len(obs_wires) != len(set(obs_wires)):
-            try:
-                rotations, diag_obs = diagonalize_qwc_pauli_words(obs)
-            except ValueError as e:
-                raise qml.QuantumFunctionError(
-                    "Only observables that are qubit-wise commuting "
-                    "Pauli words can be returned on the same wire"
-                ) from e
+    # Check for observables acting on the same wire. If present, observables must be
+    # qubit-wise commuting Pauli words. In this case, the tape is expanded with joint
+    # rotations and the observables updated to the computational basis. Note that this
+    # expansion acts on the original tape in place.
+    if tape._repeated_observables:
+        try:
+            rotations, diag_obs = diagonalize_qwc_pauli_words(tape._repeated_observables)
+        except ValueError as e:
```

Comment:
Tangent: something I've been pondering for a while - I think we rely too much on built-in python exceptions. This is more than basic input validation, so should probably be a custom exception (e.g., `GroupingException`). This way we can catch specific grouping errors with better accuracy, while allowing `ValueError` exceptions (which could be caused by other bugs/issues) to continue pass through unimpeded.

Reply:
That is a good point!